pallets/auctions/src/lib.rs

```diff
@@ -0,0 +1,1356 @@
+		/// Candle auction cannot have a reserve price
+		CandleAuctionDoesNotSupportReservePrice,
+		/// Math overflow
+		Overflow,
+		/// Error when determining the auction winner
+		ErrorDeterminingAuctionWinner,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		///
+		/// Creates a new auction for a given Auction type
+		///
+		/// - calls the create() implementation on the given Auction type
+		/// 
+		#[pallet::weight(<T as Config>::WeightInfo::create_auction())]
+		pub fn create(origin: OriginFor<T>, auction: Auction<T>) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+
+			match &auction {
+				Auction::English(auction_object) => {
+					auction_object.create(sender, &auction)?;
+				}
+				Auction::TopUp(auction_object) => {
+					auction_object.create(sender, &auction)?;
+				}
+				Auction::Candle(auction_object) => {
+					auction_object.create(sender, &auction)?;
+				}
+			}
+
+			Ok(())
+		}
+
+		///
+		/// Updates an existing auction which has not yet started
+		/// 
+		/// - calls the update() implementation on the given Auction type
+		/// 
+		#[pallet::weight(<T as Config>::WeightInfo::update_auction())]
+		pub fn update(origin: OriginFor<T>, id: T::AuctionId, updated_auction: Auction<T>) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+
+			match updated_auction {
+				Auction::English(auction_object) => {
+					auction_object.update(sender, id)?;
+				}
+				Auction::TopUp(auction_object) => {
+					auction_object.update(sender, id)?;
+				}
+				Auction::Candle(auction_object) => {
+					auction_object.update(sender, id)?;
+				}
+			}
+			Ok(())
+		}
+
+		///
+		/// Destroys an existing auction which has not yet started
+		///
+		/// - validates write action
+		/// - unfreezes NFT
+		/// - calls destroy helper function
+		///
+		#[pallet::weight(<T as Config>::WeightInfo::destroy_auction())]
+		pub fn destroy(origin: OriginFor<T>, id: T::AuctionId) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+			let auction = <Auctions<T>>::get(id).ok_or(Error::<T>::AuctionDoesNotExist)?;
+
+			match &auction {
+				Auction::English(auction_object) => {
+					Self::validate_update(sender, &auction_object.general_data)?;
+					Self::unfreeze_nft(&auction_object.general_data)?;
+					Self::handle_destroy(id)?;
+				}
+				Auction::TopUp(auction_object) => {
+					Self::validate_update(sender, &auction_object.general_data)?;
+					Self::unfreeze_nft(&auction_object.general_data)?;
+					Self::handle_destroy(id)?;
+				}
+				Auction::Candle(auction_object) => {
+					Self::validate_update(sender, &auction_object.general_data)?;
+					Self::unfreeze_nft(&auction_object.general_data)?;
+					Self::handle_destroy(id)?;
+				}
+			}
+
+			Ok(())
+		}
+
+		///
+		/// Places a bid on a running auction
+		///
+		/// - validates bid
+		/// - calls the bid() implementation on the given Auction type
+		/// - deposits BidPlaced event
+		///
+		#[pallet::weight(<T as Config>::WeightInfo::bid())]
```

Comment:
i imagine that weight for bid can be vastly different for every auction type, how do you solve this issue? is the weight the weight for most expensive bid and you're just refunding if you use cheaper auction type? or can you somehow change weight based on parameter before even extrinsic is dispatched?